
#SheCodes PROJEECT WORK :

##GROUP MEMBERS:

Priyal Gandhi

Shivani Rajput

Riya Tyagi

Dhruvita Banugaria

###MEDICAL DIAGNOSIS : Prediction of Diesease from symptoms.

Cleaning our data

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
import csv
from collections import defaultdict

disease_list = []

def return_list(disease):
    disease_list = []
    match = disease.replace('^','_').split('_')
    ctr = 1
    for group in match:
        if ctr%2==0:
            disease_list.append(group)
        ctr = ctr + 1

    return disease_list
    
with open("/content/dataset_uncleaned.csv",encoding ='latin') as csvfile:
    reader = csv.reader(csvfile)
    disease=""
    weight = 0
    disease_list = []
    dict_wt = {}
    dict_=defaultdict(list)
    for row in reader:

        if row[0]!="\xc2\xa0" and row[0]!="":
            disease = row[0]
            disease_list = return_list(disease)
            weight = row[1]

        if row[2]!="\xc2\xa0" and row[2]!="":
            symptom_list = return_list(row[2])

            for d in disease_list:
                for s in symptom_list:
                    dict_[d].append(s)
                dict_wt[d] = weight

Writing our cleaned data

In [ ]:
with open("/content/dataset_clean.csv","w") as csvfile:
    writer = csv.writer(csvfile)
    for key,values in dict_.items():
        for v in values:
            #key = str.encode(key)
            key = str.encode(key).decode('utf-8')
            #.strip()
            #v = v.encode('utf-8').strip()
            #v = str.encode(v)
            writer.writerow([key,v,dict_wt[key]])

In [ ]:
columns = ['Source','Target','Weight']

In [ ]:
data = pd.read_csv("/content/dataset_clean.csv",names=columns, encoding ="ISO-8859-1")

In [ ]:
data.head(10)

Source               Target  Weight
0  hypertensive disease           pain chest    3363
1  hypertensive disease  shortness of breath    3363
2  hypertensive disease            dizziness    3363
3  hypertensive disease             asthenia    3363
4  hypertensive disease                 fall    3363
5  hypertensive disease              syncope    3363
6  hypertensive disease              vertigo    3363
7  hypertensive disease                sweat    3363
8  hypertensive disease   sweating increased    3363
9  hypertensive disease          palpitation    3363

In [ ]:
data.to_csv("/content/dataset_clean.csv",index=False)

In [ ]:
slist = []
dlist = []
with open("/content/nodetable.csv","w") as csvfile:
    writer = csv.writer(csvfile)

    for key,values in dict_.items():
        for v in values:
            if v not in slist:
                writer.writerow([v,v,"symptom"])
                slist.append(v)
        if key not in dlist:
            writer.writerow([key,key,"disease"])
            dlist.append(key)

In [ ]:
nt_columns = ['Id','Label','Attribute']

In [ ]:
nt_data = pd.read_csv("/content/nodetable.csv",names=nt_columns, encoding ="ISO-8859-1",)

In [ ]:
nt_data.head()

Id                Label Attribute
0           pain chest           pain chest   symptom
1  shortness of breath  shortness of breath   symptom
2            dizziness            dizziness   symptom
3             asthenia             asthenia   symptom
4                 fall                 fall   symptom

In [ ]:
nt_data.to_csv("/content/nodetable.csv",index=False)

Analysing the cleaned data

In [ ]:
data = pd.read_csv("/content/dataset_clean.csv", encoding ="ISO-8859-1")

In [ ]:

data.head()

Source               Target  Weight
0  hypertensive disease           pain chest    3363
1  hypertensive disease  shortness of breath    3363
2  hypertensive disease            dizziness    3363
3  hypertensive disease             asthenia    3363
4  hypertensive disease                 fall    3363

In [ ]:
len(data['Source'].unique())

149

In [ ]:
len(data['Target'].unique())

405

In [ ]:
df = pd.DataFrame(data)

In [ ]:

df_1 = pd.get_dummies(df.Target)

In [ ]:
df_s = df['Source']

In [ ]:
df_pivoted = pd.concat([df_s,df_1], axis=1)

In [ ]:
df_pivoted.drop_duplicates(keep='first',inplace=True)

In [ ]:
df_pivoted[:5]

Source  Heberden's node  ...  worry  yellow sputum
0  hypertensive disease                0  ...      0              0
1  hypertensive disease                0  ...      0              0
2  hypertensive disease                0  ...      0              0
3  hypertensive disease                0  ...      0              0
4  hypertensive disease                0  ...      0              0

[5 rows x 405 columns]

In [ ]:
df_pivoted = df_pivoted.groupby('Source').sum()
df_pivoted = df_pivoted.reset_index()
df_pivoted[:5]

Source  ...  yellow sputum
0                     Alzheimer's disease  ...              0
1                                     HIV  ...              0
2    Pneumocystis carinii pneumonia  ...              1
3             accident cerebrovascular  ...              0
4  acquired immuno-deficiency syndrome  ...              0

[5 rows x 405 columns]

In [ ]:
cols = df_pivoted.columns

In [ ]:
df_pivoted.to_csv("/content/df_pivoted.csv")

In [ ]:

cols = cols[1:]

In [ ]:

x = df_pivoted[cols]
y = df_pivoted['Source']

Trying out our classifier to learn diseases from the symptoms.

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import train_test_split

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=42)

In [ ]:
from sklearn.linear_model import LogisticRegression
classifier= LogisticRegression()
classifier.fit(x_train,y_train)
Y_pred = classifier.predict(x_test) 

In [ ]:
mnb_tot = MultinomialNB()
mnb_tot = mnb_tot.fit(x, y)

In [ ]:
mnb_tot.score(x, y)

0.8993288590604027

Train and Test split

In [ ]:
disease_pred = mnb_tot.predict(x)

In [ ]:
disease_real = y.values

In [ ]:
for i in range(0, len(disease_real)):
    if disease_pred[i]!=disease_real[i]:
        print ('Pred: {0} Actual:{1}'.format(disease_pred[i], disease_real[i]))

Pred: HIV Actual:acquired immuno-deficiency syndrome
Pred: biliary calculus Actual:cholelithiasis
Pred: coronary arteriosclerosis Actual:coronary heart disease
Pred: depression mental Actual:depressive disorder
Pred: HIV Actual:hiv infections
Pred: carcinoma breast Actual:malignant neoplasm of breast
Pred: carcinoma of lung Actual:malignant neoplasm of lung
Pred: carcinoma prostate Actual:malignant neoplasm of prostate
Pred: carcinoma colon Actual:malignant tumor of colon
Pred: candidiasis Actual:oralcandidiasis
Pred: effusion pericardial Actual:pericardial effusion body substance
Pred: malignant neoplasms Actual:primary malignant neoplasm
Pred: sepsis (invertebrate) Actual:septicemia
Pred: sepsis (invertebrate) Actual:systemic infection
Pred: tonic-clonic epilepsy Actual:tonic-clonic seizures


These are the predicted versus actual diseases that our classifier misclassifies.


Analysis of the Manual Training Model

In [ ]:
data = pd.read_csv("/content/Training.csv")

In [ ]:
data.head()

itching  skin_rash  ...  yellow_crust_ooze         prognosis
0        1          1  ...                  0  Fungal infection
1        0          1  ...                  0  Fungal infection
2        1          0  ...                  0  Fungal infection
3        1          1  ...                  0  Fungal infection
4        1          1  ...                  0  Fungal infection

[5 rows x 133 columns]

In [ ]:
# Import PyDrive and associated libraries.
# This only needs to be done once per notebook.
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client.
# This only needs to be done once per notebook.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

# Download a file based on its file ID.
#
# A file ID looks like: laggVyWshwcyP6kEI-y_W3P8D26sz
file_id = 'REPLACE_WITH_YOUR_FILE_ID'
downloaded = drive.CreateFile({'id': file_id})
print('Downloaded content "{}"'.format(downloaded.GetContentString()))

# New section

In [ ]:

data.columns

Index(['itching', 'skin_rash', 'nodal_skin_eruptions', 'continuous_sneezing',
       'shivering', 'chills', 'joint_pain', 'stomach_pain', 'acidity',
       'ulcers_on_tongue',
       ...
       'blackheads', 'scurring', 'skin_peeling', 'silver_like_dusting',
       'small_dents_in_nails', 'inflammatory_nails', 'blister',
       'red_sore_around_nose', 'yellow_crust_ooze', 'prognosis'],
      dtype='object', length=133)

In [ ]:
len(data.columns)

133

In [ ]:
len(data['prognosis'].unique())

41

41 different type of target diseases are available in the manual training dataset.
Checking with 'Prognosis' as it is the last element.

In [ ]:
df = pd.DataFrame(data)

In [ ]:

df.head()

itching  skin_rash  ...  yellow_crust_ooze         prognosis
0        1          1  ...                  0  Fungal infection
1        0          1  ...                  0  Fungal infection
2        1          0  ...                  0  Fungal infection
3        1          1  ...                  0  Fungal infection
4        1          1  ...                  0  Fungal infection

[5 rows x 133 columns]

In [ ]:

len(df)

4920

In [ ]:
cols = df.columns

In [ ]:

cols = cols[:-1]

In [ ]:

cols

Index(['itching', 'skin_rash', 'nodal_skin_eruptions', 'continuous_sneezing',
       'shivering', 'chills', 'joint_pain', 'stomach_pain', 'acidity',
       'ulcers_on_tongue',
       ...
       'pus_filled_pimples', 'blackheads', 'scurring', 'skin_peeling',
       'silver_like_dusting', 'small_dents_in_nails', 'inflammatory_nails',
       'blister', 'red_sore_around_nose', 'yellow_crust_ooze'],
      dtype='object', length=132)

In [ ]:
len(cols)

132

We have 132 symptoms in the manual training data.

In [ ]:

x = df[cols]
y = df['prognosis']


 Classifier to learn diseases from the symptoms

In [ ]:

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=42)

In [ ]:
mnb = MultinomialNB()
mnb = mnb.fit(x_train, y_train)

In [ ]:
mnb.score(x_test, y_test)

1.0

We use the testing dataset to actually test our Multinomial Bayes model.

Multinomial Bayes is a type of Classification.

Using Confusion Matrix to show the accuracy senstivity of our model.

In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, Y_pred)

array([[37,  0,  0, ...,  0,  0,  0],
       [ 0, 42,  0, ...,  0,  0,  0],
       [ 0,  0, 42, ...,  0,  0,  0],
       ...,
       [ 0,  0,  0, ..., 41,  0,  0],
       [ 0,  0,  0, ...,  0, 40,  0],
       [ 0,  0,  0, ...,  0,  0, 44]])

In [ ]:
classifier.score(x,y)

1.0

ACCURACY = (sum of diagonal element)/(sum of all elements)

ACCURACY = 1.0

